### DISCUSSION TITLE
Writing a simple userspace program that does something useful with this driver

### DISCUSSION THREAD

**Discussion by alexflint** [NVIDIA/open-gpu-kernel-modules#530] [Q&A] [answered]
Are there any examples or documents that would help me write a simple userspace program that interacts with this driver and does something useful, without relying on CUDA / OpenGL / GLX / X11?

For example, given some machine code generated by the PTX compiler, could I load and run that code, and read the results back on the CPU? Could I do this for, say, the [simple vector addition example](https://docs.nvidia.com/cuda/ptx-compiler-api/index.html#example-simple-vector-addition) from the PTX docs?

(I realize there is no ABI guarantee with this driver, but I'm willing to target a particular driver version. I also realize that there is probably quite a lot of driver state to set up before doing something like this.)

**Comment by CarloRamponi** [comment]
Hello @alexflint ,
I have the same question here, any luck with this?

**Reply by alexflint** [reply]
Nothing yet. The best clue I found was the nvproxy in gvisor, but it's not exactly what I described in the question here. https://github.com/google/gvisor/blob/master/pkg/sentry/devices/nvproxy/nvproxy.go

**Comment by mtijanic** [comment, marked as answer]
Hi, sorry, we don't actually have any such documentation to share. It would be nice to have, but it really hasn't been a priority for us.

I can share two toy examples that might get you started though.

- [rmmsg.c](https://gist.github.com/mtijanic/32b077542b7e56dc603074115fcaea75) lets you configure the log verbosity at runtime, instead of via `NVreg_RmMsg` as explained in https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/197
  - This is a super simple client that doesn't care about GPUs at all and is just operating on driver global state.
- [perfboost.c](https://gist.github.com/mtijanic/9c129900bfba774b39914ad11b0041f6) locks your GPU to the P0 pstate.
  - This one operates on the GPU, but has a baked in assumption of only single GPU in the system

Both of them have copies of the param structures which are not ABI stable, but in practice they have not changed over the past few years so you should be able to use them with any driver version.


Hopefully this gets you started. If you have specific questions, I can answer them. And if someone feels like writing documentation, I'll gladly review.
